[SPARK-30310] [Core] Resolve missing match case in SparkUncaughtExceptionHandler and added tests

tinhto-000 opened the pull request:
### What changes were proposed in this pull request?
1) Added missing match case to SparkUncaughtExceptionHandler, so that it would not halt the process when the exception doesn't match any of the match case statements.
2) Added log message before halting process.  During debugging it wasn't obvious why the Worker process would DEAD (until we set SPARK_NO_DAEMONIZE=1) due to the shell-scripts puts the process into background and essentially absorbs the exit code.
3) Added SparkUncaughtExceptionHandlerSuite.  Basically we create a Spark exception-throwing application with SparkUncaughtExceptionHandler and then check its exit code.

### Why are the changes needed?
SPARK-30310, because the process would halt unexpectedly.

### How was this patch tested?
All unit tests (mvn test) were ran and OK.

tinhto-000 committed:
fix missing match case in SparkUncaughtExceptionHandler
added SparkUncaughtExceptionHandlerSuite

tinhto-000 committed:
put new line at end of SparkUncaughtExceptionHandlerSuite.scala

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

seanli-rallyhealth commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -48,11 +48,17 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
-      case oom: OutOfMemoryError => Runtime.getRuntime.halt(SparkExitCode.OOM)
-      case t: Throwable => Runtime.getRuntime.halt(SparkExitCode.UNCAUGHT_EXCEPTION_TWICE)
+      case oom: OutOfMemoryError =>
+        logError("Uncaught OutOfMemoryError in thread " + thread + ", process halted.", oom)
```
use string interpolation  s"Uncaught OutOfMemoryError in thread $thread, process halted."?

tinhto-000 replied:
thanks for the suggestion!  modded accordingly.

seanli-rallyhealth commented on core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala:
```diff
@@ -0,0 +1,135 @@
+    Utils.executeCommand(
+      Seq(s"$sparkHome/bin/spark-class",
+        exceptionThrower.getClass.getCanonicalName.dropRight(1), // drops the "$" at the end
+        if (exitOnUncaughtException) "true" else "false"),
```
why not exitOnUncaughtException.toString?

tinhto-000 replied:
thanks for the suggestion!  modded accordingly.
also changed to use .toBoolean for converting string to boolean.

seanli-rallyhealth commented on core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala:
```diff
@@ -0,0 +1,135 @@
+
+object OutOfMemoryErrorThrower {
+  def main(args: Array[String]): Unit = {
+    val t = new ThrowableThrowerThread(new OutOfMemoryError, if (args(0) == "true") true else false)
```
what if args is empty?

tinhto-000 replied:
thanks for the suggestion!  modded the throwers to check args.length and exit with -1 if it's empty now.

tinhto-000 committed:
use string interpolation in log messages
SparkUncaughtExceptionHandlerSuite had wrong name
use toBoolean/toString for converting string to/from boolean
exception throwers now check args.length before using args(0)

SparkQA commented:
**[Test build #4975 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4975/testReport)** for PR 26955 at commit [`724e254`](https://github.com/apache/spark/commit/724e2540a5be3c8f8067f3d3db6a21167ad97fdf).

SparkQA commented:
**[Test build #4975 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4975/testReport)** for PR 26955 at commit [`724e254`](https://github.com/apache/spark/commit/724e2540a5be3c8f8067f3d3db6a21167ad97fdf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Jenkins retest this please

SparkQA commented:
**[Test build #115817 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115817/testReport)** for PR 26955 at commit [`724e254`](https://github.com/apache/spark/commit/724e2540a5be3c8f8067f3d3db6a21167ad97fdf).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/20609/
Test PASSed.

SparkQA commented:
**[Test build #115817 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/115817/testReport)** for PR 26955 at commit [`724e254`](https://github.com/apache/spark/commit/724e2540a5be3c8f8067f3d3db6a21167ad97fdf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/115817/
Test PASSed.

HeartSaVioR commented on core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala:
```diff
@@ -0,0 +1,156 @@
+// - SparkUncaughtExceptionHandler (SparkExitCode)
+// - main() (0, or -1 when args is empty)
+
+object RuntimeExceptionThrower {
```
All objects have same implementation except which exception to create and pass. Why not consolidating into one object, and additional argument to differentiate the exception? Define a new Enumeration for this and pass it as String for additional argument.

tinhto-000 replied:
Thanks for the suggestion!   Modified as such.

HeartSaVioR commented on core/src/test/scala/org/apache/spark/util/SparkUncaughtExceptionHandlerSuite.scala:
```diff
@@ -0,0 +1,156 @@
+      Map("SPARK_TESTING" -> "1", "SPARK_HOME" -> sparkHome))
+  }
+
+  test("SPARK-30310: Test uncaught RuntimeException, exitOnUncaughtException = true") {
```
I guess Spark codebase has been dealing with this kind of case via having Seq/Map to construct test matrix, and leverage `foreach` to construct each test. Like:

```
Seq(
  ("RuntimeException", RuntimeExceptionThrower, true, SparkExitCode.UNCAUGHT_EXCEPTION),
  ...
).foreach { case (excKind, thrower, exitOnUncaughtException, expectedExitCode) =>
  test(s"SPARK-30310: Test uncaught $excKind, exitOnUncaughtException = $exitOnUncaughtException") {
    val process = getThrowerProcess(thrower, exitOnUncaughtException = exitOnUncaughtException)
    assert(process.waitFor == expectedExitCode)
  }
}
```

Having Enumeration makes this even simpler, as `excKind` and `thrower` would be consolidated into one.

HeartSaVioR replied:
And then getThrowerProcess can be inlined as it's only used in single place.

tinhto-000 replied:
Thanks for the suggestion!   Modified as such.

HeartSaVioR commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -48,11 +48,17 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
-      case oom: OutOfMemoryError => Runtime.getRuntime.halt(SparkExitCode.OOM)
-      case t: Throwable => Runtime.getRuntime.halt(SparkExitCode.UNCAUGHT_EXCEPTION_TWICE)
+      case oom: OutOfMemoryError =>
+        logError(s"Uncaught OutOfMemoryError in thread $thread, process halted.", oom)
```
According to the content of try statement as of now, I guess the case we catch some exception here might probably be the case which Spark even fails to log. We may be better to do nothing but just halt.

tinhto-000 replied:
Thanks for the comment.  

Well the reason why for the logError is because it wasn't obvious to users or devs why the worker would just disappeared as DEAD on the UI, and there was nothing in the worker log file to tell what happened.  We couldn't find out why until we set SPARK_NO_DAEMONIZE=1 and examined the exit code.

Is there any alternative to indicate the process halted unexpectedly?

HeartSaVioR replied:
Please take a look at the content of try statement - if you don't have any log message, then logError in try statement would have been failed already. I don't think we can succeed logging here and below in Throwable.

HeartSaVioR replied:
So in that case, well, I guess exit code would be only the hint.

srowen replied:
Why would logging not succeed here?
This still halts. I think it avoids masking the OOM with a MatchError.

tinhto-000 replied:
Thanks for the comment.

But the `ShutdownHookManager.inShutdown()` call in the try block could also trigger another exception though, at least in theory?

But yes I agree, if the logError triggers it in the try then doing another logError in the catch won't do any good.

tinhto-000 replied:
I agree if it were the logError in the try block that triggers another exception then doing another logError won't do any good.  But for other cases where logError still works, it would provide better debugging messages.

HeartSaVioR replied:
Oh OK. I agree ShutdownHookManager.inShutdown() could be...

My point is that logging here could defeat the effort of catch statement, as if logError throws another exception, the return code becomes out of control. If we really want to log it here, we may want to wrap with try-catch statement again.

tinhto-000 replied:
Thanks for the comment!  How about something like this?  Consider if we get there we don't really care why logError fails.

```
        try { logError(s"Uncaught OutOfMemoryError in thread $thread, process halted.", oom) }
        catch { case _: Throwable => }
        Runtime.getRuntime.halt(SparkExitCode.OOM)
```

srowen replied:
OK, I get it, though it is a corner case on a corner case - fatal error, then logging throws a fatal error. It might be overkill to further wrap it, but, I am not against it for complete defensiveness.

HeartSaVioR replied:
I'd support wrapping it, as here we control the value of return code from 50 to 52 and doesn't sound ideal to have some case breaking it.

tinhto-000 replied:
Thanks @srowen and @HeartSaVioR, I'll make that into the next commit.

HeartSaVioR commented:
cc. @gatorsmile @zsxwing

srowen commented:
Ping @tinhto-000

tinhto-000 committed:
consolidated the thrower classes into one
use Seq().foreach() to launch the test-cases

tinhto-000 committed:
add try-catch around the second logError

HeartSaVioR commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -48,11 +48,24 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
-      case oom: OutOfMemoryError => Runtime.getRuntime.halt(SparkExitCode.OOM)
-      case t: Throwable => Runtime.getRuntime.halt(SparkExitCode.UNCAUGHT_EXCEPTION_TWICE)
+      case oom: OutOfMemoryError =>
+        try { logError(s"Uncaught OutOfMemoryError in thread $thread, process halted.", oom) }
```
Thanks for the update! It's a nit, but maybe better to be consistent on style:

```
try {
...
} catch {
...
}
```

as one-liner of try statement and multi-lines or catch statement look inconsistent.

tinhto-000 replied:
I actually agree!  Will modify.

HeartSaVioR reviewed (approved):
LGTM except a nit

tinhto-000 committed:
modify the try-block to non-oneliner.

srowen commented on core/src/main/scala/org/apache/spark/util/SparkUncaughtExceptionHandler.scala:
```diff
@@ -48,11 +48,28 @@ private[spark] class SparkUncaughtExceptionHandler(val exitOnUncaughtException:
+        try {
+          logError(s"Uncaught OutOfMemoryError in thread $thread, process halted.", oom)
+        }
+        catch {
```
Nit: pull the catch up onto the previous line, here and below

tinhto-000 replied:
Done! :-)

tinhto-000 committed:
pull catch statement up one line

HeartSaVioR commented:
@srowen Would you mind if I ask about revisiting this? Maybe with "ok to test". Thanks in advance!

srowen commented:
Jenkins, test this please

SparkQA commented:
**[Test build #116422 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116422/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21211/
Test PASSed.

SparkQA commented:
**[Test build #116422 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116422/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116422/
Test FAILed.

srowen commented:
Jenkins, test this please

SparkQA commented:
**[Test build #116438 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116438/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).

HeartSaVioR commented:
It's know failure. I pinged @HyukjinKwon to kindly ask about taking a look (as I don't have a context on pyspark) but it seems to be better to disable the test first as multiple PRs are affected.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21226/
Test PASSed.

HyukjinKwon commented:
Sorry I missed pings @HeartSaVioR - I am currently stuck at some works :(.

HeartSaVioR commented:
Oh OK. What about disabling the test and file an issue then? I've submitted #27158 to disable the test.

SparkQA commented:
**[Test build #116438 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116438/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116438/
Test FAILed.

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21241/
Test PASSed.

SparkQA commented:
**[Test build #116454 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116454/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).

SparkQA commented:
**[Test build #116454 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116454/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116454/
Test FAILed.

HeartSaVioR commented:
retest this, please

srowen commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #116645 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116645/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/21424/
Test PASSed.

SparkQA commented:
**[Test build #116645 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/116645/testReport)** for PR 26955 at commit [`3f9e96f`](https://github.com/apache/spark/commit/3f9e96f6e95c9375920a6b9e8e85c03e88b5cb7b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/116645/
Test PASSed.

srowen commented:
Merged to master
